omf-python/src/colormap.rs

```diff
@@ -0,0 +1,212 @@
+pub struct PyNumberRangeDate(pub NumberRange);
+
+#[gen_stub_pymethods]
+#[pymethods]
+impl PyNumberRangeDate {
+    #[getter]
+    fn min(&self, py: Python<'_>) -> PyResult<PyObject> {
+        match &self.0 {
+            NumberRange::Date { min, .. } => Ok((*min).into_py(py)),
+            _ => Err(PyValueError::new_err(
+                "NumberRange variant is not supported",
+            )),
+        }
+    }
+    #[getter]
+    fn max(&self, py: Python<'_>) -> PyResult<PyObject> {
+        match &self.0 {
+            NumberRange::Date { max, .. } => Ok((*max).into_py(py)),
+            _ => Err(PyValueError::new_err(
+                "NumberRange variant is not supported",
+            )),
+        }
+    }
+}
+
+#[gen_stub_pyclass]
+#[pyclass(name = "NumberRangeDateTime")]
+/// Specifies the minimum and maximum values of a number colormap.
+///
+/// Values outside this range will use the color at the ends of the gradient.
+/// The variant used should match the type of the number array.
+pub struct PyNumberRangeDateTime(pub NumberRange);
+
+#[gen_stub_pymethods]
+#[pymethods]
+impl PyNumberRangeDateTime {
+    #[getter]
+    fn min(&self, py: Python<'_>) -> PyResult<PyObject> {
+        match &self.0 {
+            NumberRange::DateTime { min, .. } => Ok((*min).into_py(py)),
+            _ => Err(PyValueError::new_err(
+                "NumberRange variant is not supported",
+            )),
+        }
+    }
+    #[getter]
+    fn max(&self, py: Python<'_>) -> PyResult<PyObject> {
+        match &self.0 {
+            NumberRange::DateTime { max, .. } => Ok((*max).into_py(py)),
+            _ => Err(PyValueError::new_err(
+                "NumberRange variant is not supported",
+            )),
+        }
+    }
+}
+
+#[gen_stub_pyclass]
+#[pyclass(name = "NumberColormapContinuous")]
+/// A continuous colormap linearly samples a color gradient within a defined range.
+///
+/// A value X% of way between `min` and `max` should use the color from X% way down
+/// gradient. When that X doesn't land directly on a color use the average of
+/// the colors on either side, inverse-weighted by the distance to each.
+///
+/// Values below the minimum use the first color in the gradient array. Values above
+/// the maximum use the last.
+pub struct PyNumberColormapContinuous(pub NumberColormap);
+
+#[gen_stub_pymethods]
+#[pymethods]
+impl PyNumberColormapContinuous {
+    #[getter]
+    /// Value range.
+    fn range(&self, py: Python<'_>) -> PyResult<PyObject> {
+        match &self.0 {
+            NumberColormap::Continuous { range, .. } => match &range {
+                NumberRange::Float { .. } => Ok(PyNumberRangeFloat(*range).into_py(py)),
+                NumberRange::Integer { .. } => Ok(PyNumberRangeInteger(*range).into_py(py)),
+                NumberRange::Date { .. } => Ok(PyNumberRangeDate(*range).into_py(py)),
+                NumberRange::DateTime { .. } => Ok(PyNumberRangeDateTime(*range).into_py(py)),
```

Comment:
I'm wondering if we need these python binding objects for ranges. Do they have a use other than wrapping a min max value? 

You could instead return [range.min, range.max] and then we don't need.

Reply:
You probably could, I do wonder if date ranges are actually used in colormaps... but anyway, I can give it a go if you like?

Reply:
Yeah I think it's worth doing

Reply:
Hi @andrew-catalyst, I did try to simplify this further, by getting rid of the variant matching in the following, but then I couldn't access `range.min`/`range.max`. Let me know if there's a better way?

```rs
match &self.0 {
            NumberColormap::Continuous { range, .. } => match &range {
                NumberRange::Float { min, max } => {
                    result.set_item("min", min)?;
                    result.set_item("max", max)?;
                }
...
```

Reply:
I'll have a look

Reply:
This seems to work:
```
    /// Value range.
    fn range(&self, py: Python<'_>) -> PyResult<PyObject> {
        match self.0 {
            NumberColormap::Continuous { range, .. } => match range {
                NumberRange::Float { min, max, .. } => Ok([min, max].into_py(py)),
                NumberRange::Integer { min, max, .. } => Ok([min, max].into_py(py)),
                NumberRange::Date { min, max, .. } => Ok([min, max].into_py(py)),
                NumberRange::DateTime { min, max, .. } => Ok([min, max].into_py(py)),
            },
            _ => Err(PyValueError::new_err(
                "NumberColormap variant is not supported",
            )),
        }
    }
```

Reply:
Actually you can use `(min,max).into_py(py)` so it returns a tuple.